src/devices/machine/i82586.cpp

```diff
@@ -73,47 +234,1752 @@ device_memory_interface::space_config_vector i82586_base_device::memory_space_co
+			compute_crc(buf, length - 4, cfg_crc16()), *(u32 *)&buf[length - 4]);
+
+		// increment crc error count
+		m_space->write_word(m_scb_address + 8, m_space->read_word(m_scb_address + 8) + 1);
+
+		rfd_cs |= RFD_S_CRC;
+	}
+#endif
+
+	// TODO: alignment error (crc in misaligned frame), status bit 10
+	// TODO: increment alignment error counter
+
+	// fetch initial rbd offset from rfd
+	u16 rbd_offset = m_space->read_word(m_rfd + 6);
+
+	if (!cfg_no_src_add_ins())
+	{
+		// compute stored length (from 2 * addresses + length field)
+		int actual = cfg_address_length() * 2 + 2;
+
+		LOG("ru_execute storing %d bytes into rfd\n", actual);
+
+		// store data in rfd
+		store_bytes(m_rfd + 8, buf, actual);
+		position += actual;
+		remaining -= actual;
+	}
+
+	// store remaining bytes in receive buffers
+	while (remaining && rbd_offset != RBD_EMPTY)
+	{
+		// fetch the count and address for this buffer
+		u32 rb_address = m_space->read_dword(m_scb_base + rbd_offset + 4);
+		u16 rbd_size = m_space->read_word(m_scb_base + rbd_offset + 8);
+
+		// compute number of bytes to store in buffer
+		int actual = remaining > (rbd_size & RB_SIZE) ? (rbd_size & RB_SIZE) : remaining;
+
+		LOG("ru_execute storing %d bytes into receive buffer 0x%08x size %d\n", actual, rb_address, rbd_size & RB_SIZE);
+
+		// store data in buffer
+		store_bytes(rb_address, &buf[position], actual);
+		position += actual;
+		remaining -= actual;
+
+		// store actual count
+		m_space->write_word(m_scb_base + rbd_offset + 0, actual | RB_F | (remaining ? 0 : RB_EOF));
+
+		// check if buffers exhausted
+		if ((rbd_size & RB_EL))
+		{
+			rbd_offset = RBD_EMPTY;
+
+			if (remaining)
+			{
+				// set buffers exhausted status
+				rfd_cs |= RFD_S_BUFFER;
+
+				m_ru_state = RU_NR;
+				m_rnr = true;
+			}
+		}
+		else
+			// fetch next rbd offset
+			rbd_offset = m_space->read_word(m_scb_base + rbd_offset + 2);
+	}
+
+	if (remaining == 0 || cfg_save_bad_frames())
+		// set frame received status
+		rfd_cs |= RFD_C;
+
+	// frame received without errors
+	if (!(rfd_cs & RFD_ERROR_82586))
+	{
+		LOG("ru_execute frame received without error\n");
+
+		rfd_cs |= RFD_OK;
+	}
+	else
+		LOG("ru_execute frame received with errors status 0x%04x\n", rfd_cs);
+
+	// store status
+	m_space->write_dword(m_rfd, rfd_cs);
+
+	// if we received without error, or we're saving bad frames, advance to the next rfd
+	if ((rfd_cs & RFD_OK) || cfg_save_bad_frames())
+	{
+		if (!(rfd_cs & RFD_EL))
+		{
+			// advance to next rfd
+			m_rfd = m_scb_base + m_space->read_word(m_rfd + 4);
+
+			// store next free rbd address into rfd
+			if (rbd_offset != RBD_EMPTY)
+				m_space->write_word(m_rfd + 6, rbd_offset);
+		}
+		else
+		{
+			m_ru_state = RU_NR;
+			m_rnr = true;
+		}
+
+		// set frame received status
+		m_fr = true;
+	}
+
+	// suspend on completion
+	if (rfd_cs & RFD_S)
+	{
+		m_ru_state = RU_SUSPENDED;
+		m_rnr = true;
+	}
+
+#if VERBOSE & LOG_GENERAL
```

Comment:
trouble with putting this inside `#if` is it can still rot if people make changes and compile without logging enabled.  That's the reason for the log macros being how they are - they at least check that the expressions compile even when logging is disabled.  Same with the other stuff in `#if` blocks.